client/blockchain-service/src/handler_msp.rs

```diff
@@ -854,6 +943,123 @@ where
         }
     }
 
+    /// Verifies that all MSP bucket forest roots match the on-chain roots.
+    ///
+    /// This is a sanity check after coming out of sync to ensure mutations were
+    /// correctly applied during the sync process.
+    ///
+    /// This function creates the forest storage under a key if it doesn't exist. This is because a user could have
+    /// created a bucket during the downtime, and since the MSP didn't confirm any storage request for it, no mutations
+    /// were applied and as such the forest storage was not created previously during the initial sync.
+    ///
+    /// TODO: A bucket could have been unassigned from this MSP if a user requested to move it to another MSP.
+    /// We should handle this by deleting the bucket's forest storage and cleaning up the file storage from
+    /// the file keys of that bucket.
+    async fn verify_msp_bucket_roots(
+        &mut self,
+        block_hash: &Runtime::Hash,
+        msp_id: &ProviderId<Runtime>,
+    ) {
+        // Get all buckets managed by this MSP
+        let buckets = match self
+            .client
+            .runtime_api()
+            .query_buckets_for_msp(*block_hash, msp_id)
+        {
+            Ok(Ok(buckets)) => buckets,
+            Ok(Err(e)) => {
+                error!(target: LOG_TARGET, "Failed to query buckets for MSP during root verification: {:?}", e);
+                return;
+            }
+            Err(e) => {
+                error!(target: LOG_TARGET, "Runtime API call failed for query_buckets_for_msp: {:?}", e);
+                return;
+            }
+        };
```

Comment:
Related to my previous concern, I think it would be wise to add to the error logs that this current state is critical, since we cannot skipping the verification of the bucket forests and cannot assume a synced state.